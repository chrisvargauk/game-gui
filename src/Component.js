export class Component {
  constructor (option) {
    this.option                     = option;
    this.id                         = Date.now() + '-' + Math.random();
    this.listObjCompChild           = {};
    this.html                       = '';
    this.dom                        = document.createElement('div');
    this.ctrChild                   = -1;
    this.state                      = {};
    this.isStateUpdated             = false;
    this.dataFromParentAsStringPrev = undefined;
  }

  renderToHtmlAndDomify ( dataFromParent ) {
    // Skip if Data Passed in from Parent Comp. is the same
    // or if State has changed
    if ( this.html !== '' ) {
      const renderBecauseDataPassedInChanged = this.isRenderBecauseDataPassedInChanged ( dataFromParent );
      if ( !renderBecauseDataPassedInChanged &&
        !this.isStateUpdated
      ) {
        return false;
      }
    }

    // Render to HTML String
    this.ctrChild       = -1;
    var htmlNewRender   = this.render( dataFromParent );
    this.isStateUpdated = false;
    this.ctrChild       = -1;

    // Skip ReDomify if Comp Shallow Render looks the same
    if ( htmlNewRender === this.html) {
      return false;
    }

    this.html = htmlNewRender;

    // DOMify HTML String
    this.dom.innerHTML = this.html;

    // Do predefined bindings automatically right after DOM is ready
    this.doBind();

    // Run Life Cycle Method if defined on Comp Instance
    if  (typeof this.afterRender !== 'undefined') {
      this.afterRender();
    }

    // Inject Child Comps if any
    this.replacePlaceholderAll( this.dom );

    return true;
  }

  isRenderBecauseDataPassedInChanged ( dataFromParent ) {
    // Scenario 1)  prev === undefined && passed === undefined --> DON'T RENDER
    if ( typeof this.dataFromParentAsStringPrev === 'undefined' &&
      typeof dataFromParent                  === 'undefined'
    ) {
      return false;
    }

    // Scenario 2)  prev !== undefined && passed === undefined --> RENDER
    if ( typeof this.dataFromParentAsStringPrev !== 'undefined' &&
      typeof dataFromParent                  === 'undefined'
    ) {
      this.dataFromParentAsStringPrev = dataFromParent;
      return true;
    }

    // Scenario 3)  prev === undefined && passed !== undefined --> COMPARE
    if ( typeof this.dataFromParentAsStringPrev === 'undefined' &&
      typeof dataFromParent                  !== 'undefined'
    ) {
      const dataFromParentAsString = JSON.stringify( dataFromParent );

      // Compare / same       --> DON'T RENDER
      if ( this.dataFromParentAsStringPrev === dataFromParentAsString) {
        return false;

        // Compare / different  --> RENDER
      } else {
        this.dataFromParentAsStringPrev = dataFromParentAsString;
        return true;
      }
    }

    // Scenario 4)  prev !== undefined && passed !== undefined --> COMPARE
    if ( typeof this.dataFromParentAsStringPrev !== 'undefined' &&
      typeof dataFromParent                  !== 'undefined'
    ) {
      const dataFromParentAsString = JSON.stringify( dataFromParent );

      // Compare / same       --> DON'T RENDER
      if ( this.dataFromParentAsStringPrev === dataFromParentAsString) {
        return false;

        // Compare / different  --> RENDER
      } else {
        this.dataFromParentAsStringPrev = dataFromParentAsString;
        return true;
      }
    }

    console.error('Something unexpected happened o_0?');
    debugger;
  }

  replacePlaceholderAll () {
    // Get a list of all Placeholder Child Component
    var nodeListPlaceholder = this.dom.querySelectorAll('.comp-placeholder');

    // Iterate over all Placeholder Child Comps and replace them with Comp DOM
    for (var indexNodeListPlaceholder = 0; indexNodeListPlaceholder < nodeListPlaceholder.length; indexNodeListPlaceholder++) {
      var dPlaceholder = nodeListPlaceholder[ indexNodeListPlaceholder ];
      var compChild = this.listObjCompChild[indexNodeListPlaceholder];
      compChild.dom = dPlaceholder.insertAdjacentElement('beforebegin', compChild.dom);
      dPlaceholder.parentNode.removeChild(dPlaceholder);
    }
  };

  include ( ClassComp, dataFromParent ) {
    // If Smart Comp (Class)
    if ( Function.prototype.toString.call(ClassComp).indexOf('class') === 0 ) {
      this.ctrChild++;
      var nameComp  = ClassComp.name;
      var compChild = this.listObjCompChild[ this.ctrChild ];

      // Create new instance of Comp only if it hasn't been created yet
      if ( typeof compChild === 'undefined' ) {
        compChild = this.listObjCompChild[ this.ctrChild ] = new ClassComp( this.option );

        // Hook up scheduler
        // Note: make sure you dont bind this, you need scheduler to resolve to ui framework,
        //       and not to comp instance
        compChild.scheduleRendering = this.scheduleRendering;
        compChild.scheduleRendering( this );
      }

      compChild.renderToHtmlAndDomify( dataFromParent );

      return `<div class="comp-placeholder" type="${nameComp}" ctr-child="${this.ctrChild}"></div>`;

    // If Dumb Comp (Function)
    } else {
      return ClassComp( dataFromParent, this.option );
    }
  };

  doBind () {
    // ui-click | Do click handler bindings automatically right after rendering
    // Note: Binding happens before Child Comps are injected so Encapsulation is unharmed
    const listNodeBtn = this.dom.querySelectorAll('[ui-click]');
    for (let indexListNodeBtn = 0; indexListNodeBtn < listNodeBtn.length; indexListNodeBtn++) {
      const nodeBtn = listNodeBtn[ indexListNodeBtn ];

      const handler = nodeBtn.getAttribute('ui-click');
      if (typeof this[handler] === 'undefined') {
        console.warn(`UI: click handler called "${handler}" can't be found on the Component.`);
        continue;
      }

      nodeBtn.addEventListener('click', this[handler].bind(this), false);
    }
  }

  getState () {
    return this.state;
  };

  setState ( stateNew ) {
    this.state = stateNew;
    this.isStateUpdated = true;

    if ( typeof this.scheduleRendering !== 'undefined' ) {
      this.scheduleRendering( this );
    }
  };
}

export default Component;