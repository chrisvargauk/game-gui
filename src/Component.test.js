import Component from "./Component";

describe('Component', () => {
  class ComponentTest extends Component {
    constructor( option, config ) {
      super(  option, config  );
    }
  }

  describe('constructor', () => {

    it('Takes "option" and "config" objects as arguments at instantiation.', () => {
      const componentTest = new ComponentTest({foo: 'foo'}, {bar: 'bar'});
      expect( componentTest.option.foo ).toBe('foo');
      expect( componentTest.config.bar ).toBe('bar');
    });

    it('Arguments "option" and "config" are optional.', () => {
      let componentTest,
          error = false;
      try {
        componentTest = new ComponentTest();
      } catch ( err ) {
        error = err;
      }

      expect( error ).toBe( false );
      expect( JSON.stringify(componentTest.option) ).toBe('{}');
      expect( JSON.stringify(componentTest.config) ).toBe('{}');
    });

    describe('Comp ID is', () => {
      it('Either defined in the config (config.id),', () => {
        const componentTest = new ComponentTest(undefined, {id: 'foo'});
        expect( componentTest.id ).toBe( 'foo' );
      });

      it('or random generated if not specified in the config.', () => {
        const componentTestA = new ComponentTest();
        const componentTestB = new ComponentTest();
        expect( typeof componentTestA.id ).toBe( 'string' );
        expect( componentTestA.id !== componentTestB.id).toBe( true );
      });
    });

    it('Should have "type" prop. which is the name of the class', () => {
      const componentTest = new ComponentTest();
      expect( componentTest.type ).toBe( 'ComponentTest' );
      // note: detailed testing of "type" will be done when we test the method that produces it,
      //       here we just wanna know that this prop is present.
    });

    it('Should have "dom" prop., the DOM Element that will contain all the rendered elements of the Comp.', () => {
      const componentTest = new ComponentTest();
      expect( componentTest.dom instanceof HTMLElement ).toBe( true );
      // note: detailed testing of "dom" will be done when we test it factory method,
      //       here we just wanna know that this prop is present .
    });

    it('Should have rest of the default props set correctly', () => {
      const componentTest = new ComponentTest();
      componentTest.id = 'static-id-for-test-sake';
      expect( componentTest ).toMatchSnapshot();
    });

  });

  describe('getTypeOfComp', () => {
    it('Should return "type", which is the class name.', () => {
      const componentTest = new ComponentTest();
      expect( componentTest.type ).toBe( 'ComponentTest' );
    });

    it('Should take artifacts created by transpiler into account when "type" is returned.', () => {
      class ComponentTestCustom_ImportedCompName extends Component {};
      // Note: we mimic what the transpiler does to class names.

      const componentTestCustom = new ComponentTestCustom_ImportedCompName();

      expect( componentTestCustom.type ).toBe( 'ImportedCompName' );
    });
  });

  describe('createDomElement', () => {
    const componentTest = new ComponentTest();
    const domElement = componentTest.createDomElement( 'fooBar', 'bar' );

    it('Should return an HTML Element', () => {
      expect( domElement instanceof HTMLElement ).toBe( true );
    });

    it('Should set the Comps type as Snake Cased class name on returned HTML Element.', () => {
      expect( domElement.classList.contains('foo-bar') ).toBe( true );
    });

    it('Should set the Comps ID as "id" attribute on returned HTML Element.', () => {
      expect( domElement.getAttribute('id') ).toBe( 'bar' );
    });
  });
});