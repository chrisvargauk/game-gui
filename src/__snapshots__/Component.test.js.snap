// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Should have rest of the default props set correctly 1`] = `
ComponentTest {
  "config": Object {},
  "ctrChild": -1,
  "dataFromParentAsStringPrev": undefined,
  "dom": <div
    class="component-test"
    id="15723215432609177"
  />,
  "html": "",
  "id": "static-id-for-test-sake",
  "isStateUpdated": false,
  "listObjCompChild": Object {},
  "option": Object {},
  "state": Object {},
  "type": "ComponentTest",
}
`;

exports[`Component constructor Should have rest of the default props set correctly 1`] = `
ComponentTest {
  "config": Object {},
  "ctrChild": -1,
  "dataFromParentAsStringPrev": undefined,
  "dom": <div
    class="component-test"
    id="157232174580498610"
  />,
  "html": "",
  "id": "static-id-for-test-sake",
  "isStateUpdated": false,
  "listObjCompChild": Object {},
  "option": Object {},
  "state": Object {},
  "type": "ComponentTest",
}
`;
